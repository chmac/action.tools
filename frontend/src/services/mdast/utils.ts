import * as R from "remeda";
import dayjs from "dayjs";
import { RuleOption } from "@rschedule/core";

import { Rule } from "../../rschedule";
import {
  EVERY,
  AFTER,
  UNITS,
  DAYS,
  MONTHS,
  MONTHS_TO_NUMBER
} from "./mdast.service";

export const getNumberFromStart = (input: string): number => {
  const result = input.match(/^[\d]+/);
  if (!result || !result.length || !result[0]) {
    throw new Error("Failed to find number #1jMxvU");
  }
  const number = parseInt(result[0]);
  if (typeof number !== "number" || number <= 0) {
    throw new Error("Invalid number found #GPU80T");
  }
  return number;
};
export const isValidUnit = (unit: string): unit is dayjs.OpUnitType => {
  return UNITS.indexOf(unit) !== -1;
};

export const convertUnit = (unit: string) => {
  return unit.toLowerCase().replace(/s$/, "");
};

export const getCountAndUnitFromString = (
  input: string
): { count: number; unit: dayjs.OpUnitType } => {
  const result = input.match(/^([\d]+)([\w]+)$/);

  if (
    !result ||
    typeof result[1] !== "string" ||
    typeof result[2] !== "string"
  ) {
    throw new Error("Invalid recurrence string #LK6o80");
  }

  const [_, countString, unitResult] = result;

  const count = parseInt(countString);

  if (count <= 0) {
    throw new Error("Invalid count #E8DCQc");
  }

  const unit = convertUnit(unitResult);

  if (!isValidUnit(unit)) {
    throw new Error("Invalid unit specified #Fvk3ZI");
  }

  return {
    count,
    unit
  };
};

export const startsWithANumber = (input: string): boolean => {
  return !!input.match(/^\d/);
};

export const isDay = (input: string): boolean => {
  return DAYS.indexOf(input) !== -1;
};

export const isMonth = (input: string): boolean => {
  return MONTHS.indexOf(input) !== -1;
};
export const isValidPoint = (point: string) => {};

export const startsWith = (target: string, input: string): boolean => {
  return input.substr(0, target.length).toLowerCase() === target;
};

export const removeFromFront = (target: string, input: string): string => {
  if (!startsWith(target, input)) {
    return input;
  }
  return input.substr(target.length);
};

export const dayToRuleDay = (day: string): RuleOption.ByDayOfWeek => {
  if (!isDay(day)) {
    throw new Error("dayToRuleDay called with invalid day #TesvaG");
  }
  return day.substr(0, 2).toUpperCase() as RuleOption.ByDayOfWeek;
};

export const monthToRuleMonth = (month: string): RuleOption.ByMonthOfYear => {
  if (!isMonth(month)) {
    throw new Error("monthToRuleMonth called with invalid month #6rFg6u");
  }
  const lowerMonth = month.toLowerCase();
  return MONTHS_TO_NUMBER[lowerMonth] as RuleOption.ByMonthOfYear;
};

export const calculateNextOccurrence = (schedule: string, from: Date): Date => {
  if (startsWith(EVERY, schedule)) {
    const extendBy = removeFromFront(EVERY, schedule);
    if (startsWithANumber(extendBy)) {
      const { count, unit } = getCountAndUnitFromString(extendBy);
      return dayjs(from)
        .add(count, unit)
        .toDate();
    }

    const points = extendBy.split(",");

    const arePointsAllDays = points.every(isDay);
    const arePointsAllMonths = points.every(isMonth);
    if (!(arePointsAllDays || arePointsAllMonths)) {
      throw new Error("Invalid mixture of days and months #V0yUbX");
    }

    if (arePointsAllDays) {
      const rule = new Rule({
        frequency: "WEEKLY",
        byDayOfWeek: R.map(points, dayToRuleDay),
        start: from
      });
      const [next] = rule.occurrences({ start: from, take: 1 }).toArray();

      return new Date(next.toISOString());
    } else if (arePointsAllMonths) {
      const rule = new Rule({
        frequency: "YEARLY",
        byMonthOfYear: R.map(points, monthToRuleMonth),
        start: from
      });

      const [next] = rule.occurrences({ start: from, take: 1 }).toArray();

      return new Date(next.toISOString());
    }
  } else if (startsWith(AFTER, schedule)) {
    const extendBy = removeFromFront(AFTER, schedule);
    const { count, unit } = getCountAndUnitFromString(extendBy);
    return dayjs(from)
      .add(count, unit)
      .toDate();
  }
  throw new Error("Invalid repetition schedule #222CJy");
};
