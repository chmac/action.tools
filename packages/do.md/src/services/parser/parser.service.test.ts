import { List } from 'mdast';
import {
  complexTaskListMarkdown,
  getFirstTaskFromMdast,
  markdownToMdast,
  taskListOnlyMarkdown,
  tasksWithSecondLineMarkdown,
} from '../../__fixtures__/markdown.fixtures';
import {
  getDataFromListItem,
  getTextFromListItem,
  listToTasks,
  parseMdast,
} from './parser.service';

describe('parser', () => {
  // We test the fixtures to ensure that as we upgrade / swap the packages that
  // convert text to mdast and back, they don't introduce changes.
  describe('fixtures', () => {
    it('Fixtures remain consistent', () => {
      expect(markdownToMdast(taskListOnlyMarkdown)).toMatchSnapshot();
      expect(markdownToMdast(tasksWithSecondLineMarkdown)).toMatchSnapshot();
      expect(markdownToMdast(`- [ ] # Can we heading?`)).toMatchSnapshot();
      expect(
        markdownToMdast(`- [ ] # Will a heading fit here?  
  
  Multiple paras?`)
      ).toMatchSnapshot();
    });
  });

  describe('getTextFromListItem()', () => {
    it('Gets text from a simple task #ERa2eO', () => {
      const item = getFirstTaskFromMdast(
        markdownToMdast(`- [ ] A simple task`)
      );

      expect(getTextFromListItem(item)).toEqual('A simple task');
    });

    it('Gets text from a task with 2 lines #0yybP5', () => {
      const item = getFirstTaskFromMdast(
        markdownToMdast(`- [ ] A second line task  \nWith a second line`)
      );

      expect(getTextFromListItem(item)).toEqual(
        `A second line task\nWith a second line`
      );
    });

    it('Gets text from a task with 2 lines and inline code #6Bf3LC', () => {
      const item = getFirstTaskFromMdast(
        markdownToMdast(
          '- [ ] A second line task  \nWith a second line and a by date `by:2020-02-24`'
        )
      );

      expect(getTextFromListItem(item)).toEqual(
        `A second line task\nWith a second line and a by date`
      );
    });
  });

  describe('getDataFromListItem()', () => {
    it('Returns empty object for task with no data #szIeqo', () => {
      expect(
        getDataFromListItem(
          getFirstTaskFromMdast(
            markdownToMdast(`- [ ] A simple task without any data`)
          )
        )
      ).toEqual({});
    });

    it('Fetches data correctly #szIeqo', () => {
      expect(
        getDataFromListItem(
          getFirstTaskFromMdast(
            markdownToMdast(
              '- [ ] A simple task without any data `after:2020-02-01` `by:2020-02-24` `id:def12`'
            )
          )
        )
      ).toEqual({
        after: '2020-02-01',
        by: '2020-02-24',
        id: 'def12',
      });
    });
  });

  describe('listToTasks()', () => {
    it('Generates tasks #HetQot', () => {
      const root = markdownToMdast(taskListOnlyMarkdown);
      expect(listToTasks(root.children[0] as List)).toMatchSnapshot();
    });
  });

  describe('parseMdast()', () => {
    it('Generates sections with tasks #otm70r', () => {
      // NOTE: This snapshot is so we have a record of the input as well as the
      // expected output
      expect(markdownToMdast(complexTaskListMarkdown)).toMatchSnapshot();
      expect(
        parseMdast(markdownToMdast(complexTaskListMarkdown))
      ).toMatchSnapshot();
    });

    it('debugging #gVI0DY', () => {
      // This makes the snapshots a bit easier to see, huge mdast trees make it
      // hard to see what's going on otherwise.
      const sections = parseMdast(markdownToMdast(complexTaskListMarkdown));
      const sectionDeets = sections.map(section => ({
        id: section.id,
        heading: section.heading?.children[0],
        taskCount: section.tasks.length,
      }));
      expect(sectionDeets).toMatchSnapshot();
    });
  });
});
