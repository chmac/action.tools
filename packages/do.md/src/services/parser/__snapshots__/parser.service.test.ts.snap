// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser fixtures Fixtures remain consistent 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "checked": false,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 20,
                      "line": 1,
                      "offset": 19,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 7,
                      "line": 1,
                      "offset": 6,
                    },
                  },
                  "type": "text",
                  "value": "A simple task",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 20,
                  "line": 1,
                  "offset": 19,
                },
                "indent": Array [],
                "start": Object {
                  "column": 7,
                  "line": 1,
                  "offset": 6,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 20,
              "line": 1,
              "offset": 19,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": false,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 27,
                      "line": 2,
                      "offset": 46,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 7,
                      "line": 2,
                      "offset": 26,
                    },
                  },
                  "type": "text",
                  "value": "A second simple task",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 27,
                  "line": 2,
                  "offset": 46,
                },
                "indent": Array [],
                "start": Object {
                  "column": 7,
                  "line": 2,
                  "offset": 26,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 27,
              "line": 2,
              "offset": 46,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 20,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": false,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 27,
                      "line": 3,
                      "offset": 73,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 7,
                      "line": 3,
                      "offset": 53,
                    },
                  },
                  "type": "text",
                  "value": "One more simple task",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 27,
                  "line": 3,
                  "offset": 73,
                },
                "indent": Array [],
                "start": Object {
                  "column": 7,
                  "line": 3,
                  "offset": 53,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 27,
              "line": 3,
              "offset": 73,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 3,
              "offset": 47,
            },
          },
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": Position {
        "end": Object {
          "column": 27,
          "line": 3,
          "offset": 73,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "spread": false,
      "start": null,
      "type": "list",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 27,
      "line": 3,
      "offset": 73,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`parser fixtures Fixtures remain consistent 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "checked": false,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 19,
                      "line": 1,
                      "offset": 18,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 7,
                      "line": 1,
                      "offset": 6,
                    },
                  },
                  "type": "text",
                  "value": "A first task",
                },
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 3,
                      "line": 2,
                      "offset": 23,
                    },
                    "indent": Array [
                      3,
                    ],
                    "start": Object {
                      "column": 19,
                      "line": 1,
                      "offset": 18,
                    },
                  },
                  "type": "break",
                },
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 29,
                      "line": 2,
                      "offset": 49,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 2,
                      "offset": 23,
                    },
                  },
                  "type": "text",
                  "value": "With a second line of text",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 29,
                  "line": 2,
                  "offset": 49,
                },
                "indent": Array [
                  3,
                ],
                "start": Object {
                  "column": 7,
                  "line": 1,
                  "offset": 6,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 29,
              "line": 2,
              "offset": 49,
            },
            "indent": Array [
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": false,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "position": Position {
                    "end": Object {
                      "column": 42,
                      "line": 3,
                      "offset": 91,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 7,
                      "line": 3,
                      "offset": 56,
                    },
                  },
                  "type": "text",
                  "value": "A second task without a second line",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 42,
                  "line": 3,
                  "offset": 91,
                },
                "indent": Array [],
                "start": Object {
                  "column": 7,
                  "line": 3,
                  "offset": 56,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 42,
              "line": 3,
              "offset": 91,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 3,
              "offset": 50,
            },
          },
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": Position {
        "end": Object {
          "column": 42,
          "line": 3,
          "offset": 91,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "spread": false,
      "start": null,
      "type": "list",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 42,
      "line": 3,
      "offset": 91,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`parser listToTasks() Generates tasks #HetQot 1`] = `
Array [
  Object {
    "checked": false,
    "children": Array [
      Object {
        "children": Array [
          Object {
            "position": Position {
              "end": Object {
                "column": 20,
                "line": 1,
                "offset": 19,
              },
              "indent": Array [],
              "start": Object {
                "column": 7,
                "line": 1,
                "offset": 6,
              },
            },
            "type": "text",
            "value": "A simple task",
          },
        ],
        "position": Position {
          "end": Object {
            "column": 20,
            "line": 1,
            "offset": 19,
          },
          "indent": Array [],
          "start": Object {
            "column": 7,
            "line": 1,
            "offset": 6,
          },
        },
        "type": "paragraph",
      },
    ],
    "id": "{\\"start\\":{\\"line\\":1,\\"column\\":1,\\"offset\\":0},\\"end\\":{\\"line\\":1,\\"column\\":20,\\"offset\\":19},\\"indent\\":[]}",
    "position": Position {
      "end": Object {
        "column": 20,
        "line": 1,
        "offset": 19,
      },
      "indent": Array [],
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "spread": false,
    "taskData": Object {},
    "text": "A simple task",
    "type": "listItem",
  },
  Object {
    "checked": false,
    "children": Array [
      Object {
        "children": Array [
          Object {
            "position": Position {
              "end": Object {
                "column": 27,
                "line": 2,
                "offset": 46,
              },
              "indent": Array [],
              "start": Object {
                "column": 7,
                "line": 2,
                "offset": 26,
              },
            },
            "type": "text",
            "value": "A second simple task",
          },
        ],
        "position": Position {
          "end": Object {
            "column": 27,
            "line": 2,
            "offset": 46,
          },
          "indent": Array [],
          "start": Object {
            "column": 7,
            "line": 2,
            "offset": 26,
          },
        },
        "type": "paragraph",
      },
    ],
    "id": "{\\"start\\":{\\"line\\":2,\\"column\\":1,\\"offset\\":20},\\"end\\":{\\"line\\":2,\\"column\\":27,\\"offset\\":46},\\"indent\\":[]}",
    "position": Position {
      "end": Object {
        "column": 27,
        "line": 2,
        "offset": 46,
      },
      "indent": Array [],
      "start": Object {
        "column": 1,
        "line": 2,
        "offset": 20,
      },
    },
    "spread": false,
    "taskData": Object {},
    "text": "A second simple task",
    "type": "listItem",
  },
  Object {
    "checked": false,
    "children": Array [
      Object {
        "children": Array [
          Object {
            "position": Position {
              "end": Object {
                "column": 27,
                "line": 3,
                "offset": 73,
              },
              "indent": Array [],
              "start": Object {
                "column": 7,
                "line": 3,
                "offset": 53,
              },
            },
            "type": "text",
            "value": "One more simple task",
          },
        ],
        "position": Position {
          "end": Object {
            "column": 27,
            "line": 3,
            "offset": 73,
          },
          "indent": Array [],
          "start": Object {
            "column": 7,
            "line": 3,
            "offset": 53,
          },
        },
        "type": "paragraph",
      },
    ],
    "id": "{\\"start\\":{\\"line\\":3,\\"column\\":1,\\"offset\\":47},\\"end\\":{\\"line\\":3,\\"column\\":27,\\"offset\\":73},\\"indent\\":[]}",
    "position": Position {
      "end": Object {
        "column": 27,
        "line": 3,
        "offset": 73,
      },
      "indent": Array [],
      "start": Object {
        "column": 1,
        "line": 3,
        "offset": 47,
      },
    },
    "spread": false,
    "taskData": Object {},
    "text": "One more simple task",
    "type": "listItem",
  },
]
`;

exports[`parser parseMdast() Generates sections with tasks #otm70r 1`] = `
Array [
  Object {
    "contents": Array [],
    "depth": 0,
    "heading": undefined,
    "sequence": 0,
    "tasks": Array [],
  },
  Object {
    "contents": Array [],
    "depth": 0,
    "heading": Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "First heading",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 16,
          "line": 1,
          "offset": 15,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "heading",
    },
    "sequence": 1,
    "tasks": Array [
      Object {
        "checked": false,
        "children": Array [
          Object {
            "children": Array [
              Object {
                "position": Position {
                  "end": Object {
                    "column": 23,
                    "line": 3,
                    "offset": 39,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 7,
                    "line": 3,
                    "offset": 23,
                  },
                },
                "type": "text",
                "value": "A top level task",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 3,
                    "line": 4,
                    "offset": 44,
                  },
                  "indent": Array [
                    3,
                  ],
                  "start": Object {
                    "column": 23,
                    "line": 3,
                    "offset": 39,
                  },
                },
                "type": "break",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 21,
                    "line": 4,
                    "offset": 62,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 3,
                    "line": 4,
                    "offset": 44,
                  },
                },
                "type": "inlineCode",
                "value": "after:2020-02-23",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 22,
                    "line": 4,
                    "offset": 63,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 21,
                    "line": 4,
                    "offset": 62,
                  },
                },
                "type": "text",
                "value": " ",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 37,
                    "line": 4,
                    "offset": 78,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 22,
                    "line": 4,
                    "offset": 63,
                  },
                },
                "type": "inlineCode",
                "value": "by:2020-02-24",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 38,
                    "line": 4,
                    "offset": 79,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 37,
                    "line": 4,
                    "offset": 78,
                  },
                },
                "type": "text",
                "value": " ",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 49,
                    "line": 4,
                    "offset": 90,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 38,
                    "line": 4,
                    "offset": 79,
                  },
                },
                "type": "inlineCode",
                "value": "id:abc123",
              },
            ],
            "position": Position {
              "end": Object {
                "column": 49,
                "line": 4,
                "offset": 90,
              },
              "indent": Array [
                3,
              ],
              "start": Object {
                "column": 7,
                "line": 3,
                "offset": 23,
              },
            },
            "type": "paragraph",
          },
        ],
        "id": "abc123",
        "position": Position {
          "end": Object {
            "column": 49,
            "line": 4,
            "offset": 90,
          },
          "indent": Array [
            1,
          ],
          "start": Object {
            "column": 1,
            "line": 3,
            "offset": 17,
          },
        },
        "spread": false,
        "taskData": Object {
          "after": "2020-02-23",
          "by": "2020-02-24",
          "id": "abc123",
        },
        "text": "A top level task",
        "type": "listItem",
      },
      Object {
        "checked": false,
        "children": Array [
          Object {
            "children": Array [
              Object {
                "position": Position {
                  "end": Object {
                    "column": 20,
                    "line": 5,
                    "offset": 110,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 7,
                    "line": 5,
                    "offset": 97,
                  },
                },
                "type": "text",
                "value": "A second task",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 3,
                    "line": 6,
                    "offset": 115,
                  },
                  "indent": Array [
                    3,
                  ],
                  "start": Object {
                    "column": 20,
                    "line": 5,
                    "offset": 110,
                  },
                },
                "type": "break",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 21,
                    "line": 6,
                    "offset": 133,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 3,
                    "line": 6,
                    "offset": 115,
                  },
                },
                "type": "inlineCode",
                "value": "after:2020-01-01",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 22,
                    "line": 6,
                    "offset": 134,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 21,
                    "line": 6,
                    "offset": 133,
                  },
                },
                "type": "text",
                "value": " ",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 37,
                    "line": 6,
                    "offset": 149,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 22,
                    "line": 6,
                    "offset": 134,
                  },
                },
                "type": "inlineCode",
                "value": "by:2020-02-25",
              },
            ],
            "position": Position {
              "end": Object {
                "column": 37,
                "line": 6,
                "offset": 149,
              },
              "indent": Array [
                3,
              ],
              "start": Object {
                "column": 7,
                "line": 5,
                "offset": 97,
              },
            },
            "type": "paragraph",
          },
        ],
        "id": "{\\"start\\":{\\"line\\":5,\\"column\\":1,\\"offset\\":91},\\"end\\":{\\"line\\":6,\\"column\\":37,\\"offset\\":149},\\"indent\\":[1]}",
        "position": Position {
          "end": Object {
            "column": 37,
            "line": 6,
            "offset": 149,
          },
          "indent": Array [
            1,
          ],
          "start": Object {
            "column": 1,
            "line": 5,
            "offset": 91,
          },
        },
        "spread": false,
        "taskData": Object {
          "after": "2020-01-01",
          "by": "2020-02-25",
        },
        "text": "A second task",
        "type": "listItem",
      },
    ],
  },
  Object {
    "contents": Array [
      Object {
        "children": Array [
          Object {
            "position": Position {
              "end": Object {
                "column": 36,
                "line": 10,
                "offset": 202,
              },
              "indent": Array [],
              "start": Object {
                "column": 1,
                "line": 10,
                "offset": 167,
              },
            },
            "type": "text",
            "value": "Some text under the second heading.",
          },
        ],
        "position": Position {
          "end": Object {
            "column": 36,
            "line": 10,
            "offset": 202,
          },
          "indent": Array [],
          "start": Object {
            "column": 1,
            "line": 10,
            "offset": 167,
          },
        },
        "type": "paragraph",
      },
    ],
    "depth": 0,
    "heading": Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 15,
              "line": 8,
              "offset": 165,
            },
            "indent": Array [],
            "start": Object {
              "column": 4,
              "line": 8,
              "offset": 154,
            },
          },
          "type": "text",
          "value": "Sub heading",
        },
      ],
      "depth": 2,
      "position": Position {
        "end": Object {
          "column": 15,
          "line": 8,
          "offset": 165,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 8,
          "offset": 151,
        },
      },
      "type": "heading",
    },
    "sequence": 2,
    "tasks": Array [
      Object {
        "checked": false,
        "children": Array [
          Object {
            "children": Array [
              Object {
                "position": Position {
                  "end": Object {
                    "column": 17,
                    "line": 12,
                    "offset": 220,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 7,
                    "line": 12,
                    "offset": 210,
                  },
                },
                "type": "text",
                "value": "Third task",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 3,
                    "line": 13,
                    "offset": 225,
                  },
                  "indent": Array [
                    3,
                  ],
                  "start": Object {
                    "column": 17,
                    "line": 12,
                    "offset": 220,
                  },
                },
                "type": "break",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 21,
                    "line": 13,
                    "offset": 243,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 3,
                    "line": 13,
                    "offset": 225,
                  },
                },
                "type": "inlineCode",
                "value": "after:2020-02-23",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 22,
                    "line": 13,
                    "offset": 244,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 21,
                    "line": 13,
                    "offset": 243,
                  },
                },
                "type": "text",
                "value": " ",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 37,
                    "line": 13,
                    "offset": 259,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 22,
                    "line": 13,
                    "offset": 244,
                  },
                },
                "type": "inlineCode",
                "value": "by:2020-02-25",
              },
            ],
            "position": Position {
              "end": Object {
                "column": 37,
                "line": 13,
                "offset": 259,
              },
              "indent": Array [
                3,
              ],
              "start": Object {
                "column": 7,
                "line": 12,
                "offset": 210,
              },
            },
            "type": "paragraph",
          },
        ],
        "id": "{\\"start\\":{\\"line\\":12,\\"column\\":1,\\"offset\\":204},\\"end\\":{\\"line\\":13,\\"column\\":37,\\"offset\\":259},\\"indent\\":[1]}",
        "position": Position {
          "end": Object {
            "column": 37,
            "line": 13,
            "offset": 259,
          },
          "indent": Array [
            1,
          ],
          "start": Object {
            "column": 1,
            "line": 12,
            "offset": 204,
          },
        },
        "spread": false,
        "taskData": Object {
          "after": "2020-02-23",
          "by": "2020-02-25",
        },
        "text": "Third task",
        "type": "listItem",
      },
      Object {
        "checked": false,
        "children": Array [
          Object {
            "children": Array [
              Object {
                "position": Position {
                  "end": Object {
                    "column": 30,
                    "line": 14,
                    "offset": 289,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 7,
                    "line": 14,
                    "offset": 266,
                  },
                },
                "type": "text",
                "value": "Fourth task on the list",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 3,
                    "line": 15,
                    "offset": 294,
                  },
                  "indent": Array [
                    3,
                  ],
                  "start": Object {
                    "column": 30,
                    "line": 14,
                    "offset": 289,
                  },
                },
                "type": "break",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 21,
                    "line": 15,
                    "offset": 312,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 3,
                    "line": 15,
                    "offset": 294,
                  },
                },
                "type": "inlineCode",
                "value": "after:2020-01-01",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 22,
                    "line": 15,
                    "offset": 313,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 21,
                    "line": 15,
                    "offset": 312,
                  },
                },
                "type": "text",
                "value": " ",
              },
              Object {
                "position": Position {
                  "end": Object {
                    "column": 37,
                    "line": 15,
                    "offset": 328,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 22,
                    "line": 15,
                    "offset": 313,
                  },
                },
                "type": "inlineCode",
                "value": "by:2020-02-24",
              },
            ],
            "position": Position {
              "end": Object {
                "column": 37,
                "line": 15,
                "offset": 328,
              },
              "indent": Array [
                3,
              ],
              "start": Object {
                "column": 7,
                "line": 14,
                "offset": 266,
              },
            },
            "type": "paragraph",
          },
          Object {
            "children": Array [
              Object {
                "checked": false,
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "position": Position {
                          "end": Object {
                            "column": 26,
                            "line": 16,
                            "offset": 354,
                          },
                          "indent": Array [],
                          "start": Object {
                            "column": 9,
                            "line": 16,
                            "offset": 337,
                          },
                        },
                        "type": "text",
                        "value": "Nested task first",
                      },
                    ],
                    "position": Position {
                      "end": Object {
                        "column": 26,
                        "line": 16,
                        "offset": 354,
                      },
                      "indent": Array [],
                      "start": Object {
                        "column": 9,
                        "line": 16,
                        "offset": 337,
                      },
                    },
                    "type": "paragraph",
                  },
                ],
                "position": Position {
                  "end": Object {
                    "column": 26,
                    "line": 16,
                    "offset": 354,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 3,
                    "line": 16,
                    "offset": 331,
                  },
                },
                "spread": false,
                "type": "listItem",
              },
              Object {
                "checked": false,
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "position": Position {
                          "end": Object {
                            "column": 27,
                            "line": 17,
                            "offset": 381,
                          },
                          "indent": Array [],
                          "start": Object {
                            "column": 9,
                            "line": 17,
                            "offset": 363,
                          },
                        },
                        "type": "text",
                        "value": "Second nested task",
                      },
                    ],
                    "position": Position {
                      "end": Object {
                        "column": 27,
                        "line": 17,
                        "offset": 381,
                      },
                      "indent": Array [],
                      "start": Object {
                        "column": 9,
                        "line": 17,
                        "offset": 363,
                      },
                    },
                    "type": "paragraph",
                  },
                ],
                "position": Position {
                  "end": Object {
                    "column": 27,
                    "line": 17,
                    "offset": 381,
                  },
                  "indent": Array [],
                  "start": Object {
                    "column": 3,
                    "line": 17,
                    "offset": 357,
                  },
                },
                "spread": false,
                "type": "listItem",
              },
            ],
            "ordered": false,
            "position": Position {
              "end": Object {
                "column": 27,
                "line": 17,
                "offset": 381,
              },
              "indent": Array [
                3,
              ],
              "start": Object {
                "column": 3,
                "line": 16,
                "offset": 331,
              },
            },
            "spread": false,
            "start": null,
            "type": "list",
          },
        ],
        "id": "{\\"start\\":{\\"line\\":14,\\"column\\":1,\\"offset\\":260},\\"end\\":{\\"line\\":17,\\"column\\":27,\\"offset\\":381},\\"indent\\":[1,1,1]}",
        "position": Position {
          "end": Object {
            "column": 27,
            "line": 17,
            "offset": 381,
          },
          "indent": Array [
            1,
            1,
            1,
          ],
          "start": Object {
            "column": 1,
            "line": 14,
            "offset": 260,
          },
        },
        "spread": false,
        "taskData": Object {
          "after": "2020-01-01",
          "by": "2020-02-24",
        },
        "text": "Fourth task on the list",
        "type": "listItem",
      },
    ],
  },
]
`;
